torch/csrc/jit/tensorexpr/loopnest.cpp

```diff
@@ -940,13 +940,10 @@ Stmt* LoopNest::insertAllocFree(Stmt* stmt) {
   // Insert allocations and frees for temporary buffers in the innermost
   // possible scope.
   for (const Buf* buf : intermediate_bufs_) {
-    const Expr* flat_size = new IntImm(1);
-    for (auto& d : buf->dims()) {
-      flat_size = new Mul(flat_size, d);
-    }
-    flat_size = IRSimplifier::simplify(flat_size);
-    Stmt* alloc = new Allocate(buf->base_handle(), buf->dtype(), {flat_size});
-    Stmt* free = new Free(buf->base_handle());
+    const Expr* flat_size = buf_flattening_helper(buf);
+    Stmt* alloc =
+        new Allocate(new Buf(buf->base_handle(), {flat_size}, buf->dtype()));
+    Stmt* free = new Free(buf);
     Block* alloc_block = findLowestContainingBlock(uses.at(buf));
```

Comment:
FYI, when using `buf_new` instead of `buf` for this line, there are lots of failing cpp reduction tests. Didn't investigate on it though.

Reply:
We probably need to rewrite all uses of `buf` with `buf_new` before doing that. We created a new buffer, but there might be statements still referring to the old one (it wasn't the case before this change since the buf stayed the same). We probably could do that using `Substitute`, but probably it needs to be expanded to support Buf-Buf mappings:
https://github.com/pytorch/pytorch/blob/16cfe970e0dd3c7eada27ef533ed566c55850020/torch/csrc/jit/tensorexpr/eval.h#L225